src/ParallelAlgorithms/Amr/Criteria/Persson.cpp

```diff
@@ -0,0 +1,116 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ParallelAlgorithms/Amr/Criteria/Persson.hpp"
+
+#include <array>
+#include <cstddef>
+
+#include "DataStructures/ApplyMatrices.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "Domain/Amr/Flag.hpp"
+#include "NumericalAlgorithms/Spectral/Filtering.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Numeric.hpp"
+#include "Utilities/SetNumberOfGridPoints.hpp"
+
+namespace amr::Criteria {
+
+template <size_t Dim>
+double persson_smoothness_indicator(
+    const gsl::not_null<DataVector*> filtered_component_buffer,
+    const DataVector& tensor_component, const Mesh<Dim>& mesh,
+    const size_t dimension, const size_t num_highest_modes) {
+  // Zero out the lowest modes in the given dimension
+  static const Matrix identity{};
+  auto matrices = make_array<Dim>(std::cref(identity));
+  gsl::at(matrices, dimension) = Spectral::filtering::zero_lowest_modes(
+      mesh.slice_through(dimension),
+      mesh.extents(dimension) - num_highest_modes);
+  set_number_of_grid_points(filtered_component_buffer,
+                            mesh.number_of_grid_points());
+  apply_matrices(filtered_component_buffer, matrices, tensor_component,
+                 mesh.extents());
+  // Take the L2 norm over all grid points
+  return sqrt(alg::accumulate(square(*filtered_component_buffer), 0.) /
```

Comment:
I'm worried about expressions and template parameters here. Why not use Blaze L2 norm?

Reply:
Blaze's L2 norm doesn't divide by the number of grid points, and I found the docs confusing. The docs say "computes the _squared_ L1/2/3/4 norm" (https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Operations#!norms). I found it clearer to do this explicit sum-divide-sqrt.

Reply:
Okay, but the performance here is unclear. Just use the Blaze one.

Reply:
You mean `sqrt(square(blaze::l2Norm(*buffer)) / num_points)`?

Reply:
Why not `blaze::l2Norm(*buffer) / sqrt(static_cast<double>(num_points));`?

Reply:
or maybe: `(1.0 / sqrt(static_cast<double>(num_points))) * blaze::l2Norm(*buffer)`. :shrug:

Reply:
Ok